### DISCUSSION TITLE
rate limit

### DISCUSSION THREAD

**Discussion by wu-clan** [fastapi-practices/fastapi_best_architecture#70] [Ideas]
Currently known existing solutions：

1. https://github.com/abersheeran/asgi-ratelimit
2. https://github.com/long2ice/fastapi-limiter
3. https://github.com/laurentS/slowapi
4. https://github.com/Tert0/fastapi-framework/blob/develop/fastapi_framework/rate_limit.py

Make the trade-off here

**Comment by wu-clan** [comment]
Hi, @downdawn 

What do you think?

**Reply by downdawn** [reply]
slowapi is adapted from flask-limiter and is still being maintained. Although there are some problems, it should be available

**Reply by wu-clan** [reply]
I think they're all in maintenance.

My opinion is：[fastapi-limiter](https://github.com/long2ice/fastapi-limiter) > [asgi-ratelimit](https://github.com/abersheeran/asgi-ratelimit) > [slowapi](https://github.com/laurentS/slowapi)

Rationale:
- The author of `fastapi-limiter` is the same as asyncmy, tortoise-orm, in addition, the internal code is very simple, and supports websocket
- `asgi-ratelimit` is recommended by Starlette as middleware in the official documentation, it allows to define rules centrally 
- `rate_limit` I rather reject, he has to ask for request parameters, which doesn't look fastapi, I was a django user before and it asks for request parameters

**Reply by downdawn** [reply]
I'm hesitating between fastapi-limiter and slowapi because I'm more familiar with flask-limiter (which has many great features). However, after carefully reviewing the limitations of slowapi, I think I'll go with fastapi-limiter.

**Reply by wu-clan** [reply]
Agree to implement.

**Comment by wu-clan** [comment]
please review #72

**Comment by wu-clan** [comment]
[fastapi-limiter](https://github.com/long2ice/fastapi-limiter) 0.2.0 发生重大变更， 不如自己实现，相关 PR #1072 已合并
